src/os_auth/auth.c

```diff
@@ -164,62 +166,88 @@ w_err_t w_auth_parse_data(const char* buf, char *response,const char *authpass,
         }
 
     }
-    else if(!config.flags.use_source_ip) {
-        // use_source-ip = 0 and no -I argument in agent
-        snprintf(ip, IPSIZE, "any");
+    else{
+        buf--;
+        if(!config.flags.use_source_ip) {
+            // use_source-ip = 0 and no -I argument in agent
+            snprintf(ip, IPSIZE, "any");
+        }
+    }
+
+    /* Check for key hash when the agent already has one*/
+    char key_hash_token[2] = "K:";
+    if(strncmp(++buf, key_hash_token, 2)==0)
+    {
+        os_calloc(1, sizeof(os_sha1), *key_hash);
+        char format[15] = {0};
+        sprintf(format, " K:\'%%%ld[^\']\"", sizeof(os_sha1));
+        sscanf(buf, format, *key_hash);
     }
-    // else -> agent IP is already on ip
 
     return OS_SUCCESS;
 }
 
-w_err_t w_auth_validate_data (char *response, const char *ip, const char *agentname, const char *groups){
-    /* Validate the group(s) name(s) */
-    int index = 0;
-    char *id_exist = NULL;
+w_err_t w_auth_replace_agent(keyentry *key,
```

Comment:
Not sure if the name **replace** really represents the method behavior, considering it performs some checks and then removes the agent.

Reply:
All right. In fact, I can't propose an alternative right now.